plugins/wacom-i2c/fu-wac-device.c

```diff
@@ -0,0 +1,478 @@
+		g_set_error_literal (error,
+				     G_IO_ERROR,
+				     G_IO_ERROR_FAILED,
+				     "failed to send messages");
+		return FALSE;
+	}
+
+	/* first two bytes are length of the report */
+	memcpy (data, buf + 2, datasz);
+	return TRUE;
+}
+
+gboolean
+fu_wac_device_get_hid_desc (FuWacDevice *self, FuWacHidDescriptor *desc, GError **error)
+{
+	FuWacDevicePrivate *priv = GET_PRIVATE (self);
+	gchar cmd[] = {HID_DESC_REGISTER, 0x00};
+	struct i2c_msg msgs[] = {
+		{
+			.addr = priv->i2c_addr,
+			.flags = 0,
+			.len = sizeof(cmd),
+			.buf = (guint8 *)cmd,
+		},
+		{
+			.addr = priv->i2c_addr,
+			.flags = I2C_M_RD,
+			.len = sizeof(FuWacHidDescriptor),
+			.buf = (guint8 *) desc,
+		},
+	};
+	struct i2c_rdwr_ioctl_data packets = {
+		.msgs = msgs,
+		.nmsgs = 2,
+	};
+	if (ioctl (priv->fd, I2C_RDWR, &packets) < 0) {
+		g_set_error_literal (error,
+				     G_IO_ERROR,
+				     G_IO_ERROR_FAILED,
+				     "failed to send messages");
+		return FALSE;
+	}
+	return TRUE;
+}
+
+gboolean
+fu_wac_device_cmd (FuWacDevice *self,
+		   const guint8 *cmd, guint cmdsz,
+		   guint8 *rsp, guint rspsz,
+		   gulong delay_us, FuWacDeviceCmdFlags flags,
+		   GError **error)
+{
+	if (!fu_wac_device_set_feature (self,
+					FU_WAC_DEVICE_BL_REPORT_ID_CMD,
+					cmd, cmdsz, error)) {
+		g_prefix_error (error, "failed to send: ");
+		return FALSE;
+	}
+	if (delay_us > 0)
+		g_usleep (delay_us);
+	if (!fu_wac_device_get_feature (self,
+					FU_WAC_DEVICE_BL_REPORT_ID_RSP,
+					rsp, rspsz, error)) {
+		g_prefix_error (error, "failed to receive: ");
+		return FALSE;
+	}
+	if (!fu_wac_common_check_reply (cmd, rsp, error))
+		return FALSE;
+
+	/* wait for the command to complete */
+	if (flags & FU_WAC_DEVICE_CMD_FLAG_POLL_ON_WAITING &&
+	    rsp[RTRN_RSP] == FU_WAC_DEVICE_RC_WAITING) {
+		for (guint i = 0; i < FU_WAC_DEVICE_CMD_RETRIES; i++) {
+			if (delay_us > 0)
+				g_usleep (delay_us);
+			if (!fu_wac_device_get_feature (self,
+							FU_WAC_DEVICE_BL_REPORT_ID_RSP,
+							rsp, sizeof(rsp), error))
+				return FALSE;
+			if (!fu_wac_common_check_reply (cmd, rsp, error))
+				return FALSE;
+			if (rsp[RTRN_RSP] != FU_WAC_DEVICE_RC_WAITING)
+				break;
+		}
+	}
+	return fu_wac_common_rc_set_error (rsp[RTRN_RSP], error);
+}
+
+static gboolean
+fu_wac_device_set_quirk_kv (FuDevice *device,
+			    const gchar *key,
+			    const gchar *value,
+			    GError **error)
+{
+	FuWacDevice *self = FU_WAC_DEVICE (device);
+	FuWacDevicePrivate *priv = GET_PRIVATE (self);
+	if (g_strcmp0 (key, "I2cAddr") == 0) {
+		priv->i2c_addr = fu_common_strtoull (value);
+		return TRUE;
+	}
+	if (g_strcmp0 (key, "FlashBlockSize") == 0) {
+		priv->flash_block_size = fu_common_strtoull (value);
+		return TRUE;
+	}
+	if (g_strcmp0 (key, "FlashBaseAddr") == 0) {
+		priv->flash_base_addr = fu_common_strtoull (value);
+		return TRUE;
+	}
+	if (g_strcmp0 (key, "FlashSize") == 0) {
+		priv->flash_size = fu_common_strtoull (value);
+		return TRUE;
```

Comment:
Don't forget to document these in the `README.md` like we have for all the other plugins now